src/tracks/ui/BrushHandle.cpp

```diff
@@ -0,0 +1,567 @@
+   mSampleCountUpperBound = floor( pTrack->GetEndTime() * rate + 0.5);
+}
+
+BrushHandle::~BrushHandle()
+{
+}
+
+namespace {
+   // Is the distance between A and B less than D?
+   template < class A, class B, class DIST > bool within(A a, B b, DIST d)
+   {
+      return (a > b - d) && (a < b + d);
+   }
+
+   inline double findMaxRatio(double center, double rate)
+   {
+      const double minFrequency = 1.0;
+      const double maxFrequency = (rate / 2.0);
+      const double frequency =
+            std::min(maxFrequency,
+                     std::max(minFrequency, center));
+      return
+            std::min(frequency / minFrequency, maxFrequency / frequency);
+   }
+}
+
+void BrushHandle::Enter(bool, AudacityProject *project)
+{
+
+}
+
+bool BrushHandle::Escape(AudacityProject *project)
+{
+   return false;
+}
+
+// Add or remove data accroding to the ctrl key press
+void BrushHandle::HandleTimeFreqData(long long ll_sc, wxInt64 freq) {
+   // Ignore the mouse dragging outside valid area
+   // We should also check for the available freq. range that is visible to user
+   if(ll_sc < mSampleCountLowerBound || ll_sc > mSampleCountUpperBound ||
+      freq < mFreqLowerBound || freq > mFreqUpperBound)
+      return;
+
+   if(mbCtrlDown)
+      mpSpectralData->removeTimeFreqData(ll_sc, freq);
+   else
+      mpSpectralData->addTimeFreqData(ll_sc, freq);
+}
+
+UIHandle::Result BrushHandle::Click
+      (const TrackPanelMouseEvent &evt, AudacityProject *pProject)
+{
+   using namespace RefreshCode;
+
+   const auto pView = mpView.lock();
+   if ( !pView )
+      return Cancelled;
+
+   wxMouseEvent &event = evt.event;
+   const auto sTrack = TrackList::Get( *pProject ).Lock( FindTrack() );
+   const auto pTrack = sTrack.get();
+   const WaveTrack *const wt =
+         static_cast<const WaveTrack*>(pTrack);
+   auto &trackPanel = TrackPanel::Get( *pProject );
+   auto &viewInfo = ViewInfo::Get( *pProject );
+
+   mMostRecentX = event.m_x;
+   mMostRecentY = event.m_y;
+
+   return RefreshAll;
+}
+
+UIHandle::Result BrushHandle::Drag
+      (const TrackPanelMouseEvent &evt, AudacityProject *pProject)
+{
+   using namespace RefreshCode;
+
+   const auto pView = mpView.lock();
+   if ( !pView )
+      return Cancelled;
+
+   wxMouseEvent &event = evt.event;
+   const auto sTrack = TrackList::Get( *pProject ).Lock( FindTrack() );
+   const auto pTrack = sTrack.get();
+   WaveTrack *const wt =
+         static_cast<WaveTrack*>(pTrack);
+   auto &trackPanel = TrackPanel::Get( *pProject );
+   auto &viewInfo = ViewInfo::Get( *pProject );
+
+   int x = mAutoScrolling ? mMostRecentX : event.m_x;
+   int y = mAutoScrolling ? mMostRecentY : event.m_y;
+   mMostRecentX = x;
+   mMostRecentY = y;
+
+   // Clip the coordinates
+   // TODO: Find ways to access the ClipParameters (for the mid)
+   int x1 = std::max(mRect.x + 10, std::min(x, mRect.x + mRect.width - 20));
+   int y1 = std::max(mRect.y + 10, std::min(y, mRect.y + mRect.height - 10));
```

Comment:
What is the reason for the clipping of coordinates here?

I think this is a wrong solution for the absense of array bounds checking somewhere else in the program.